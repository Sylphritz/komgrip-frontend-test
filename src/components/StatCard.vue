<template>
  <div class="card">
    <div class="card-body p-3">
      <div class="row">
        <div class="col-8">
          <div class="numbers">
            <p class="text-sm mb-0 text-capitalize font-weight-bold">{{ title }}</p>
            <h5 class="font-weight-bolder mb-0">{{ formattedPrice }}</h5>
            <span
              class="text-sm font-weight-bolder"
              :class="{
                'text-success': parseFloat(changePercent) > 0,
                'text-danger': parseFloat(changePercent) < 0
              }"
              >{{ percentage }}%</span
            >
          </div>
        </div>
        <div class="col-4 text-end">
          <div class="icon icon-shape bg-gradient-primary shadow text-center border-radius-md">
            <i class="ni ni-money-coins text-lg opacity-10" aria-hidden="true"></i>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue'

const props = defineProps({
  title: {
    type: String,
    required: true
  },
  price: {
    type: String,
    required: true
  },
  changePercent: {
    type: String,
    required: true
  }
})

const percentage = computed(() => {
  const rawNumber = parseFloat(props.changePercent)

  return rawNumber < 0 ? rawNumber.toFixed(2).slice(1) : rawNumber.toFixed(2)
})
const formattedPrice = computed(() =>
  parseFloat(props.price).toLocaleString('en-US', {
    useGrouping: true,
    style: 'currency',
    currency: 'USD'
  })
)
</script>
