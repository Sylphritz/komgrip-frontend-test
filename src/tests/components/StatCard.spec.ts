import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import StatCard from '@/components/StatCard.vue'

describe('StatCard', () => {
  it('renders successfully and matches the snapshot', () => {
    const wrapper = mount(StatCard, {
      shallow: true,
      props: {
        title: 'Bitcoin',
        price: '123456.789',
        changePercent: '1.23'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders the displayed values in the correct format', () => {
    const wrapper = mount(StatCard, {
      shallow: true,
      props: {
        title: 'Bitcoin',
        price: '123456.789',
        changePercent: '1.23'
      }
    })

    expect(wrapper.find('p').element.textContent).toBe('Bitcoin')
    expect(wrapper.find('h5').element.textContent).toBe('$123,456.79')
    expect(wrapper.find('span').element.textContent).toBe('1.23%')
  })

  it('displays the change percentage in green when the value is positive', () => {
    const wrapper = mount(StatCard, {
      shallow: true,
      props: {
        title: 'Bitcoin',
        price: '123456.789',
        changePercent: '1.23'
      }
    })

    expect(wrapper.find('span').element.textContent).toBe('1.23%')
    expect(wrapper.find('span').element.classList.contains('text-success')).toBe(true)
  })

  it('displays the change percentage in red when the value is negative and no minus signs', () => {
    const wrapper = mount(StatCard, {
      shallow: true,
      props: {
        title: 'Bitcoin',
        price: '123456.789',
        changePercent: '-1.23'
      }
    })

    expect(wrapper.find('span').element.textContent).toBe('1.23%')
    expect(wrapper.find('span').element.classList.contains('text-danger')).toBe(true)
  })
})
