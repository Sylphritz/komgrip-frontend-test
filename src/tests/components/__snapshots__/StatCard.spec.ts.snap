// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatCard > renders successfully and matches the snapshot 1`] = `
<div
  class="card"
>
  <div
    class="card-body p-3"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <div
          class="numbers"
        >
          <p
            class="text-sm mb-0 text-capitalize font-weight-bold"
          >
            Bitcoin
          </p>
          <h5
            class="font-weight-bolder mb-0"
          >
            $123,456.79
          </h5>
          <span
            class="text-sm font-weight-bolder text-success"
          >
            1.23%
          </span>
        </div>
      </div>
      <div
        class="col-4 text-end"
      >
        <div
          class="icon icon-shape bg-gradient-primary shadow text-center border-radius-md"
        >
          <i
            aria-hidden="true"
            class="ni ni-money-coins text-lg opacity-10"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
